callgraph/callgraph.go

```diff
@@ -85,29 +86,52 @@ func New(root *ssa.Function, srcFns ...*ssa.Function) (*Graph, error) {
 					switch instrt := instr.(type) {
 					case *ssa.Call:
 						// debugf("found call instr")
-						fn, ok := instrt.Call.Value.(*ssa.Function)
-						if ok {
-							err := g.AddFunction(fn, allFns)
-							if err != nil {
-								return fmt.Errorf("failed to add src function %v from block instr: %w", fn, err)
+						var instrCall *ssa.Function
+
+						// Handle the case where the function calls a
+						// named function (*ssa.Function), and the case
+						// where the function calls an anonymous
+						// function (*ssa.MakeClosure).
+						switch callt := instrt.Call.Value.(type) {
+						case *ssa.Function:
+							// debugf("found call instr to function")
+							instrCall = callt
+						case *ssa.MakeClosure:
+							// debugf("found call instr to closure")
+							switch calltFn := callt.Fn.(type) {
+							case *ssa.Function:
+								instrCall = calltFn
 							}
 						}
 
+						// If we could not determine the function being
+						// called, skip this instruction.
+						if instrCall == nil {
+							continue
+						}
+
+						AddEdge(g.CreateNode(fn), instrt, g.CreateNode(instrCall))
+
+						err := g.AddFunction(instrCall, allFns)
+						if err != nil {
+							return fmt.Errorf("failed to add function %v from block instr: %w", instrCall, err)
+						}
+
 						// attempt to link function arguments that are functions
 						for a := 0; a < len(instrt.Call.Args); a++ {
 							arg := instrt.Call.Args[a]
 							switch argt := arg.(type) {
 							case *ssa.Function:
 								// TODO: check if edge already exists?
-								AddEdge(g.CreateNode(fn), instrt, g.CreateNode(argt))
+								AddEdge(g.CreateNode(instrCall), instrt, g.CreateNode(argt))
 							case *ssa.MakeClosure:
 								switch argtFn := argt.Fn.(type) {
 								case *ssa.Function:
-									AddEdge(g.CreateNode(fn), instrt, g.CreateNode(argtFn))
+									AddEdge(g.CreateNode(instrCall), instrt, g.CreateNode(argtFn))
 
 									// Assumes the anonymous functions are called.
 									for _, anFn := range argtFn.AnonFuncs {
-										AddEdge(g.CreateNode(fn), instrt, g.CreateNode(anFn))
+										AddEdge(g.CreateNode(instrCall), instrt, g.CreateNode(anFn))
```

Comment:
This ended up causing the subtle bug described in detail in https://github.com/picatz/taint/pull/5#issuecomment-1418309211